unified_planning/grpc/server.py

```diff
@@ -0,0 +1,165 @@
+    _ports: Set[int]
+    _lock = threading.Lock()
+
+    def __init__(
+        self,
+        host: str = "localhost",
+        port: Optional[int] = None,
+        override: bool = False,
+        timeout: Optional[float] = 0.5,
+    ) -> None:
+        """GRPC Planner Definition
+
+        :param host: Host address, defaults to "localhost"
+        :type host: str, optional
+        :param port: Port, defaults to None
+        :type port: Optional[int], optional
+        :param override: Override the creation of new client, defaults to False
+        :type override: bool, optional
+        :param timeout: Timeout in seconds, defaults to 0.5
+        :type timeout: Optional[float], optional
+        :raises UPException: If the gRPC server is not available or accessible
+        """
+        self._host = host
+        self._port = port
+        self._override = override
+        self._timeout_sec = timeout
+
+        self._writer = ProtobufWriter()
+        self._reader = ProtobufReader()
+
+        # Setup channel
+        self._channel = grpc.insecure_channel(
+            f"{self._host}:{self._port}",
+            options=(
+                ("grpc.enable_http_proxy", 0),
+                ("grpc.so_reuseport", 0),
+            ),
+        )
+        if not self._grpc_server_on(self._channel):
+            raise UPException(
+                "The GRPC server is not available on port {}".format(self._port)
+            )
+
+        self._planner = grpc_api.UnifiedPlanningStub(self._channel)
+
+    def __new__(cls, **kwargs) -> "GRPCPlanner":
+        """Create a thread-safe singleton instance of the planner
+
+        Modes:
+         - If parameters are provided,
+                - If the port is available, create a new client
+                - If the port is already in use, return the existing client
+         - If no parameters are provided, use default.
+        """
+        port = kwargs.get("port", None)
+
+        if (port, cls) not in cls._instances:
+            with cls._lock:
+                if cls not in cls._instances:
+                    cls._ports.add(port)
+                    cls._instances[(port, cls)] = super().__new__(cls)
+                    return cls._instances[(port, cls)]
+        elif kwargs.get("override", False):
+            return super().__new__(cls)
+        return cls._instances[(port, cls)]
+
+    def __del__(self) -> None:
+        """Delete the planner instance"""
+        self._channel.close()
+
+        for instance in self._instances:
+            if instance[1] == self:
+                self._instances.pop(instance)
+                break
+
+    def _solve(
+        self,
+        problem: "up.model.AbstractProblem",
+        callback: Optional[
+            Callable[["up.engines.results.PlanGenerationResult"], None]
+        ] = None,
+        timeout: Optional[float] = None,
+        output_stream: Optional[IO[str]] = None,
+    ) -> "up.engines.results.PlanGenerationResult":
```

Comment:
The signature of this method is changed. Now we have a dedicated OM for anytime planning and we removed the callback from this method. We also added a new parameter to specify a custom heuristic.

Have a look at the OneshotPlannerMixin class for more details.